### DISCUSSION TITLE
Display freezes when saving images to SD card

### DISCUSSION THREAD

**Discussion by embedded-kiddie** [lovyan03/LovyanGFX#616] [Q&A] [answered]
First of all, thank you for providing us with such a great library.

I am currently working on building a project that will display the images from a camera on an LCD display and save the images to an SD card when I touch the screen.

Using Adafruit's [ST7789 driver](https://github.com/adafruit/Adafruit-ST7735-Library "adafruit/Adafruit-ST7735-Library: This is a library for the Adafruit 1.8&quot; SPI display http://www.adafruit.com/products/358  and http://www.adafruit.com/products/618") and [GFX library](https://github.com/adafruit/Adafruit-GFX-Library "adafruit/Adafruit-GFX-Library: Adafruit GFX graphics core Arduino library, this is the &#39;core&#39; class that all our other graphics libraries derive from"), as well as the [XPT2046 touchscreen library](https://github.com/PaulStoffregen/XPT2046_Touchscreen "PaulStoffregen/XPT2046_Touchscreen: Touchscreen Arduino Library for XPT2046 Touch Controller Chip"), I have confirmed that basic functions work.

So I changed the graphics library to LovyanGFX (I recently ran [benchmarks of the major graphics libraries](https://github.com/embedded-kiddie/Arduino-XIAO-ESP32/tree/main/GFXbenchmark) and I found LovyanGFX is the fastest in every category!), and now I'm facing an issue where the display and touchscreen stops working when accessing the SD card.

I think this issue may be related to [this comment](https://github.com/lovyan03/LovyanGFX/discussions/370#discussioncomment-5297561 "re-init displays · lovyan03/LovyanGFX · Discussion #370") by @lovyan03:

> Just to be sure, is there an SD card (TF card) connected on the same SPI bus?
> If an SD card is present and you have not gone through the procedure to change the SD to SPI mode, the SD card may malfunction and prevent communication with the display.

To simplify this issue, I ran the [SavePNG](https://github.com/lovyan03/LovyanGFX/tree/master/examples/Standard/SavePNG "LovyanGFX/examples/Standard/SavePNG at master · lovyan03/LovyanGFX"), but it stopped after displaying `PNG save test`.

![PXL_20240912_074254464](https://github.com/user-attachments/assets/d3b329c3-ea44-4c64-b0d7-cee284398593)

And when I tried inserting `SPI.begin(...)` just before `do {...} while (...);` as shown below, a white square PNG image was saved (I think this means the save failed), but also the message `PNG save success.` was not displayed.

```c++
  // Initialize SD card interface
  SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  do {
    SD.end();
    delay(1000);

    SD.begin(SDCARD_SS_PIN, SDCARD_SPI, SPI_READ_FREQUENCY);

  } while (!saveToSD());
```

Please help me to solve this issue.
Thanks in advance.

----------

Here is the information about my development environment.

### Development environment

- [Arduino IDE 2.3.2](https://www.arduino.cc/en/software "Software｜Arduino")
- [Seeed Studio XIAO ESP32S3](https://wiki.seeedstudio.com/xiao_esp32s3_getting_started/ "Getting Started with Seeed Studio XIAO ESP32S3 (Sense)｜Seeed Studio Wiki")
  - ESP32 board package: [espressif/arduino-esp32](https://github.com/espressif/arduino-esp32 "espressif/arduino-esp32: Arduino core for the ESP32") ([3.0.4](https://github.com/espressif/arduino-esp32/releases/tag/3.0.4 "Release Arduino Release v3.0.4 based on ESP-IDF v5.1.4+ · espressif/arduino-esp32"))
  - PSRAM: Disable
- Display: ST7789
  - Interface: SPI
  - SPI Frequency: 80MHz

<details>

<summary>My custom <code>LGFX_XIAO_ESP32S3_ST7789.hpp</code></summary>

### My custom `LGFX_XIAO_ESP32S3_ST7789.hpp`

```c++
#pragma once

#define TFT_WIDTH  240 // ST7789 240 x 240 and 240 x 320
#define TFT_HEIGHT 320 // ST7789 240 x 320

// Seeed Studio XIAO ESP32-S3
#define TFT_SCLK      D8
#define TFT_MISO      D9
#define TFT_MOSI      D10
#define TFT_CS        D2
#define TFT_RST       D1
#define TFT_DC        D0
#define TOUCH_CS      D3
#define TOUCH_IRQ     D7
#define SD_CS         D6
#define SPI_MODE      SPI_MODE3

#define SPI_FREQUENCY       80000000  // 80 MHz
#define SPI_READ_FREQUENCY  25000000
#define SPI_TOUCH_FREQUENCY   250000 // 250 KHz

#define LGFX_USE_V1

#include <LovyanGFX.hpp>

// ESP32でLovyanGFXを独自設定で利用する場合の設定例

/// 独自の設定を行うクラスを、LGFX_Deviceから派生して作成します。
class LGFX : public lgfx::LGFX_Device
{
// 接続するパネルの型にあったインスタンスを用意します。
lgfx::Panel_ST7789      _panel_instance;


// パネルを接続するバスの種類にあったインスタンスを用意します。
lgfx::Bus_SPI       _bus_instance;   // SPIバスのインスタンス

// タッチスクリーンの型にあったインスタンスを用意します。(必要なければ削除)
lgfx::Touch_XPT2046          _touch_instance;

public:

  // コンストラクタを作成し、ここで各種設定を行います。
  // クラス名を変更した場合はコンストラクタも同じ名前を指定してください。
  LGFX(void)
  {
    { // バス制御の設定を行います。
      auto cfg = _bus_instance.config();      // バス設定用の構造体を取得します。

      // SPIバスの設定
      cfg.spi_host    = SPI3_HOST;            // 使用するSPIを選択
      cfg.spi_mode    = SPI_MODE;             // SPI通信モードを設定 (0 ~ 3)
      cfg.freq_write  = SPI_FREQUENCY;        // 送信時のSPIクロック
      cfg.freq_read   = SPI_READ_FREQUENCY;   // 受信時のSPIクロック
      cfg.spi_3wire   = false;                // 受信をMOSIピンで行う場合はtrueを設定
      cfg.use_lock    = true;                 // トランザクションロックを使用する場合はtrueを設定
      cfg.dma_channel = SPI_DMA_CH_AUTO;      // 使用するDMAチャンネルを設定
      cfg.pin_sclk    = TFT_SCLK;             // SPIのSCLKピン番号を設定
      cfg.pin_mosi    = TFT_MOSI;             // SPIのMOSIピン番号を設定
      cfg.pin_miso    = TFT_MISO;             // SPIのMISOピン番号を設定 (-1 = disable)
      cfg.pin_dc      = TFT_DC;               // SPIのD/Cピン番号を設定  (-1 = disable)

      _bus_instance.config(cfg);              // 設定値をバスに反映します。
      _panel_instance.setBus(&_bus_instance); // バスをパネルにセットします。
    }

    { // 表示パネル制御の設定を行います。
      auto cfg = _panel_instance.config();    // 表示パネル設定用の構造体を取得します。

      cfg.pin_cs           = TFT_CS;          // CSが接続されているピン番号   (-1 = disable)
      cfg.pin_rst          = TFT_RST;         // RSTが接続されているピン番号  (-1 = disable)
      cfg.pin_busy         = -1;              // BUSYが接続されているピン番号 (-1 = disable)

      cfg.panel_width      = TFT_WIDTH;       // 実際に表示可能な幅
      cfg.panel_height     = TFT_HEIGHT;      // 実際に表示可能な高さ
      cfg.offset_x         = 0;               // パネルのX方向オフセット量
      cfg.offset_y         = 0;               // パネルのY方向オフセット量
      cfg.offset_rotation  = 0;               // 回転方向の値のオフセット 0~7 (4~7は上下反転)
      cfg.dummy_read_pixel = 8;               // ピクセル読出し前のダミーリードのビット数
      cfg.dummy_read_bits  = 1;               // ピクセル以外のデータ読出し前のダミーリードのビット数
      cfg.readable         =  true;           // データ読出しが可能な場合 trueに設定
      cfg.invert           = false;           // パネルの明暗が反転してしまう場合 trueに設定
      cfg.rgb_order        = false;           // パネルの赤と青が入れ替わってしまう場合 trueに設定
      cfg.dlen_16bit       = false;           // 16bitパラレルやSPIでデータ長を16bit単位で送信するパネルの場合 trueに設定
      cfg.bus_shared       = true;            // SDカードとバスを共有している場合 trueに設定(drawJpgFile等でバス制御を行います)

      _panel_instance.config(cfg);
    }

    { // タッチスクリーン制御の設定を行います。（必要なければ削除）
      auto cfg = _touch_instance.config();

      cfg.x_min      = 0;                     // タッチスクリーンから得られる最小のX値(生の値)
      cfg.x_max      = TFT_WIDTH-1;           // タッチスクリーンから得られる最大のX値(生の値)
      cfg.y_min      = 0;                     // タッチスクリーンから得られる最小のY値(生の値)
      cfg.y_max      = TFT_HEIGHT-1;          // タッチスクリーンから得られる最大のY値(生の値)
      cfg.pin_int    = TOUCH_IRQ;             // INTが接続されているピン番号
      cfg.bus_shared = true;                  // 画面と共通のバスを使用している場合 trueを設定
      cfg.offset_rotation = 0;                // 表示とタッチの向きのが一致しない場合の調整 0~7の値で設定

// SPI接続の場合
      cfg.spi_host = SPI3_HOST;               // 使用するSPIを選択 (HSPI_HOST or VSPI_HOST)
      cfg.freq = SPI_TOUCH_FREQUENCY;         // SPIクロックを設定
      cfg.pin_sclk = TFT_SCLK;                // SCLKが接続されているピン番号
      cfg.pin_mosi = TFT_MOSI;                // MOSIが接続されているピン番号
      cfg.pin_miso = TFT_MISO;                // MISOが接続されているピン番号
      cfg.pin_cs   = TOUCH_CS;                //   CSが接続されているピン番号

      _touch_instance.config(cfg);
      _panel_instance.setTouch(&_touch_instance);  // タッチスクリーンをパネルにセットします。
    }

    setPanel(&_panel_instance); // 使用するパネルをセットします。
  }
};
```

</details>

<details>

<summary>Modified <code>SavePNG.ino</code></summary>

### Modified `SavePNG.ino`

```c++
#if defined (ARDUINO_WIO_TERMINAL)
  #include <Seeed_FS.h>
  #include <SD/Seeed_SD.h>
#else
  #include <SD.h>
  #include <SPIFFS.h>
#endif

//#include <LovyanGFX.hpp>
#include "LGFX_XIAO_ESP32S3_ST7789.hpp"
static LGFX lcd;

static constexpr char filename[] = "/lovyangfx_test.png";

#ifndef SDCARD_SS_PIN
#define SDCARD_SS_PIN SD_CS
#endif

#ifndef SDCARD_SPI
#define SDCARD_SPI SPI
#endif

bool saveToSD(void)
{
  std::size_t dlen;
  std::uint8_t* png = (std::uint8_t*)lcd.createPng(&dlen, 0, 0, 128, 128);
  if (!png)
  {
    Serial.print("error:createPng\n");
    return false;
  }

  Serial.print("success:createPng\n");

  bool result = false;
  File file = SD.open(filename, "w");
  if (file)
  {
    file.write((std::uint8_t*)png, dlen);
    file.close();
    result = true;
  }
  else
  {
    Serial.print("error:file open failure\n");
  }

  free(png);
  return result;
}

void setup(void)
{
  Serial.begin(115200);

  lcd.init();
  lcd.setColorDepth(24);
  lcd.setColor(TFT_WHITE);

  lcd.startWrite();
  lcd.setAddrWindow(0, 0, lcd.width(), lcd.height());
  for (int y = 0; y < lcd.height(); ++y)
  {
    for (int x = 0; x < lcd.width(); ++x)
    {
      lcd.writeColor( lcd.color888(x << 1, x + y, y << 1), 1);
    }
  }
  lcd.print("PNG save test\n");
  lcd.endWrite();

  // Initialize SD card interface
  SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  do {
    SD.end();
    delay(1000);

    SD.begin(SDCARD_SS_PIN, SDCARD_SPI, SPI_READ_FREQUENCY);

  } while (!saveToSD());


  Serial.println("PNG save success.");
  lcd.print("PNG save success.");
}

void loop(void)
{
  lcd.drawPngFile(SD, filename, random(-20,20), random(-20, 20));
}
```

</details>

### Wiring of my camera system

[LMX90640-XIAO-ESP32-2.4 (Ver.2)](https://app.cirkitdesigner.com/project/837fd6ec-a7d8-4381-a41f-4b953adefee0 "Cirkit Designer")

**Comment by embedded-kiddie** [comment]
There is one more thing that bothered me.

When compiling, the following warning appears:

```
/Users/myname/Documents/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/Bus_SPI.cpp:26: warning: "REG_SPI_BASE" redefined
   26 |   #define REG_SPI_BASE(i)   (DR_REG_SPI1_BASE + (((i)>1) ? (((i)* 0x1000) + 0x20000) : (((~(i)) & 1)* 0x1000 )))
      | 
In file included from /Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/esp_hw_support/include/esp_memory_utils.h:12,
                 from /Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/freertos/FreeRTOS-Kernel/portable/xtensa/include/freertos/portmacro.h:79,
                 from /Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/freertos/FreeRTOS-Kernel/include/freertos/portable.h:59,
                 from /Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/freertos/FreeRTOS-Kernel/include/freertos/FreeRTOS.h:71,
                 from /Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/driver/include/esp_private/spi_common_internal.h:13,
                 from /Users/myname/Documents/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/Bus_SPI.hpp:36,
                 from /Users/myname/Documents/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/Bus_SPI.cpp:21:
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/soc/esp32s3/include/soc/soc.h:37: note: this is the location of the previous definition
   37 | #define REG_SPI_BASE(i)         (((i)>=2) ? (DR_REG_SPI2_BASE + (i-2) * 0x1000) : (0))    // GPSPI2 and GPSPI3
      | 
In file included from /Users/myname/Documents/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32s3/Bus_RGB.cpp:29:
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/gdma_ll.h: In function 'void gdma_ll_rx_enable_interrupt(gdma_dev_t*, uint32_t, uint32_t, bool)':
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/gdma_ll.h:82:46: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
   82 |         dev->channel[channel].in.int_ena.val |= mask;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/gdma_ll.h:84:46: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
   84 |         dev->channel[channel].in.int_ena.val &= ~mask;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/gdma_ll.h: In function 'void gdma_ll_tx_enable_interrupt(gdma_dev_t*, uint32_t, uint32_t, bool)':
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/gdma_ll.h:330:47: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
  330 |         dev->channel[channel].out.int_ena.val |= mask;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/gdma_ll.h:332:47: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
  332 |         dev->channel[channel].out.int_ena.val &= ~mask;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
In file included from /Users/myname/Documents/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32s3/Bus_RGB.cpp:32:
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/lcd_ll.h: In function 'void lcd_ll_enable_interrupt(lcd_cam_dev_t*, uint32_t, bool)':
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/lcd_ll.h:558:33: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
  558 |         dev->lc_dma_int_ena.val |= mask & 0x03;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/Users/myname/Library/Arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-b6b4727c58/esp32s3/include/hal/esp32s3/include/hal/lcd_ll.h:560:33: warning: compound assignment with 'volatile'-qualified left operand is deprecated [-Wvolatile]
  560 |         dev->lc_dma_int_ena.val &= ~(mask & 0x03);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
```

I think something is displayed correctly, so I don't know if that's a problem or not 🤔
Thanks.

**Comment by tobozo** [comment]
hi, thanks for your feedback :+1: 

the (shared) bus is already started by LovyanGFX so I wonder if you really need SPI.begin :thinking: 

>   SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);



maybe try this instead:

```cpp

void setup()
{
  Serial.begin(115200);

  lcd.init();
  lcd.setColorDepth(24);
  lcd.setColor(TFT_WHITE);

  lcd.startWrite();
  lcd.setAddrWindow(0, 0, lcd.width(), lcd.height());
  for (int y = 0; y < lcd.height(); ++y)
  {
    for (int x = 0; x < lcd.width(); ++x)
    {
      lcd.writeColor( lcd.color888(x << 1, x + y, y << 1), 1);
    }
  }
  lcd.print("PNG save test\n");
  lcd.endWrite();

  // no need to re-initialize SPI, comment this out
  // SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  // assume the SD is already inserted and avoid the debug hell of { do ... while()  }
  if( SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY ) ) 
  {
    if( saveToSD() )
    {
      Serial.println("success !!");
      lcd.drawPngFile(SD, filename, random(-20,20), random(-20, 20));
    }
    else
      Serial.println("Failed to save to SD");
  } 
  else 
    Serial.println("Failed to start SD");
}

void loop()
{
   // leave empty
}




```

**Reply by embedded-kiddie** [reply]
Hi @tobozo , thank you for your comment.

> the (shared) bus is already started by LovyanGFX so I wonder if you really need SPI.begin 🤔
> 
> ```SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);```

Yes I know, but if I commented out that line, the "do ~ while()" loop continues forever.
Or I saw just `Failed to start SD` when I executed the code you attached.

It means that `SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY )` returns always `false`.

Edit: ex) Inserting some debugging code, Using JTAG debugger, Observations with oscilloscope or logic analyzer, Inserting a pull-up register, ... (I only have oscilloscope and DMM 😅

Any help on how to debug this issue would be greatly appreciated.
Thanks.

**Reply by tobozo** [reply]
I have no clue how to debug with an oscilloscope :hand_over_mouth: 

so I assume that without using the display, this would work, right?

```cpp
SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);
SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY )
```

if so, have you tried moving the SPI.begin() before tft.init() is called and see what happens?

**Reply by embedded-kiddie** [reply]
> if so, have you tried moving the SPI.begin() before tft.init() is called and see what happens?

I already do that in my application code. But to make things simple, I ran the following tests:

## TEST 1

The following works good.

```c++
void setup()
{
  Serial.begin(115200);

  // Initialize SPI
  SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  if( SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY ) ) {
    Serial.println("success !!");
    writeFile(SD, "/hello.txt", "Hello ");
    appendFile(SD, "/hello.txt", "World!\n");
    readFile(SD, "/hello.txt");
    listDir(SD, "/", 0);
  }
  else {
    Serial.println("Failed to save to SD");
  }
}

void loop()
{
   // leave empty
}
```

The functions `writeFile()`, `appendFile()`, `readFile()`, `listDir()` are from [libraries/SD/examples/SD_Test/SD_Test.ino](https://github.com/espressif/arduino-esp32/blob/master/libraries/SD/examples/SD_Test/SD_Test.ino "arduino-esp32/libraries/SD/examples/SD_Test/SD_Test.ino at master · espressif/arduino-esp32").

And I saw the followings in the serial monitor.

```
success !!
Writing file: /hello.txt
File written
Appending to file: /hello.txt
Message appended
Reading file: /hello.txt
Read from file: Hello World!
Listing directory: /
  FILE: hello.txt  SIZE: 13
```

## TEST 2

Add displaying something on LCD, but SD functions are commented out. Also works good.
 
```c++
void setup()
{
  Serial.begin(115200);

  // Initialize SPI
//SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  lcd.init();
  lcd.fillScreen(0x0000); // black
/*
  if( SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY ) ) {
    Serial.println("success !!");
    writeFile(SD, "/hello.txt", "Hello ");
    appendFile(SD, "/hello.txt", "World!\n");
    readFile(SD, "/hello.txt");
    listDir(SD, "/", 0);
  }
  else {
    Serial.println("Failed to save to SD");
  }
*/
}

void loop()
{
  lcd.setCursor(0, 0);
  lcd.setTextColor(0xFFFF); // white
  lcd.print("Hello World!");

  Serial.println("Hello World!");
  delay(1000);
}
```

## TEST3 

Restore commented out code. I saw `Hello World!` on LCD but `Failed to save to SD` in serial monitor.

## TEST 4

Change the order of initialization functions, ie:

```c++
void setup()
{
  Serial.begin(115200);

  lcd.init();
  lcd.fillScreen(0x0000); // black

  // Initialize SPI
  SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  if( SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY ) ) {
...
```

Then I could not see `Hello World!` on LCD, but saw `Failed to save to SD`.

## Conclusion

- `lcd.init()` prevents SPI for SD.
- `SPI.begin()` for SD prevents LCD working.

The situation is as described in the first thread.
Umm...

**Reply by tobozo** [reply]
some ideas:

- include <SD.h> from `LGFX_XIAO_ESP32S3_ST7789.hpp`, before including <LovyanGFX.hpp> 
- set SPI_FREQUENCY for both LCD and SD to 25MHz 
- temporarily disable touch (comment out the block in lgfx config) 
- use  `SD.begin(SD_CS)` instead of `SD.begin( SD_CS, SPI, SPI_READ_FREQUENCY )`

**Reply by embedded-kiddie** [reply]
Hi @tobozo ,
Thank you for your ideas!

> include <SD.h> from `LGFX_XIAO_ESP32S3_ST7789.hpp`, before including <LovyanGFX.hpp>

That is the original setting in [SavePNG.ino](https://github.com/lovyan03/LovyanGFX/tree/master/examples/Standard/SavePNG "LovyanGFX/examples/Standard/SavePNG at master · lovyan03/LovyanGFX") used in my testing.

And unfortunately, all your ideas (and even some combinations of those) could not solve my issue :face_exhaling:

Here is my current `SavePNG.ino` (no png saving, but just text saving only 😉):

<details>

<summary>My current <code>SavePNG.ino</code></summary>

```c++
#if defined (ARDUINO_WIO_TERMINAL)
  #include <Seeed_FS.h>
  #include <SD/Seeed_SD.h>
#else
  #include <SD.h>
  #include <SPIFFS.h>
#endif

//#include <LovyanGFX.hpp>
#include "LGFX_XIAO_ESP32S3_ST7789.hpp"
static LGFX lcd;

/* Functions from Espressif SD library
 * https://github.com/espressif/arduino-esp32/tree/master/libraries/SD/examples/SD_Test
 */
void listDir(fs::FS &fs, const char *dirname, uint8_t levels) {
  Serial.printf("Listing directory: %s\n", dirname);

  File root = fs.open(dirname);
  if (!root) {
    Serial.println("Failed to open directory");
    return;
  }
  if (!root.isDirectory()) {
    Serial.println("Not a directory");
    return;
  }

  File file = root.openNextFile();
  while (file) {
    if (file.isDirectory()) {
      Serial.print("  DIR : ");
      Serial.println(file.name());
      if (levels) {
        listDir(fs, file.path(), levels - 1);
      }
    } else {
      Serial.print("  FILE: ");
      Serial.print(file.name());
      Serial.print("  SIZE: ");
      Serial.println(file.size());
    }
    file = root.openNextFile();
  }
}

void readFile(fs::FS &fs, const char *path) {
  Serial.printf("Reading file: %s\n", path);

  File file = fs.open(path);
  if (!file) {
    Serial.println("Failed to open file " + String(path) + " for reading");
    return;
  }

  Serial.print("Read from file: ");
  while (file.available()) {
    Serial.write(file.read());
  }
  file.close();
}

void writeFile(fs::FS &fs, const char *path, const char *message) {
  Serial.printf("Writing file: %s\n", path);

  File file = fs.open(path, FILE_WRITE);
  if (!file) {
    Serial.println("Failed to open file " + String(path) + " for writing");
    return;
  }
  if (file.print(message)) {
    Serial.println("File written");
  } else {
    Serial.println("Write failed");
  }
  file.close();
}

void appendFile(fs::FS &fs, const char *path, const char *message) {
  Serial.printf("Appending to file: %s\n", path);

  File file = fs.open(path, FILE_APPEND);
  if (!file) {
    Serial.println("Failed to open file " + String(path) + " for appending");
    return;
  }
  if (file.print(message)) {
    Serial.println("Message appended");
  } else {
    Serial.println("Append failed");
  }
  file.close();
}

void setup()
{
  Serial.begin(115200);

  lcd.init();
  lcd.fillScreen(0x0000); // black

  // Re-initialize SPI for SD
//SPI.begin(TFT_SCLK, TFT_MISO, TFT_MOSI, SD_CS);

  if( SD.begin( SD_CS/*, SPI, SPI_READ_FREQUENCY*/ ) ) {
    Serial.println("success !!");
    writeFile(SD, "/hello.txt", "Hello ");
    appendFile(SD, "/hello.txt", "World!\n");
    readFile(SD, "/hello.txt");
    listDir(SD, "/", 0);
  }
  else {
    Serial.println("Failed to save to SD");
  }
}

void loop()
{
  lcd.setCursor(0, 0);
  lcd.setTextColor(0xFFFF); // white
  lcd.print("Hello World!");

  Serial.println("Hello World!");
  delay(1000);
}
```

</details>

In `LGFX_XIAO_ESP32S3_ST7789.hpp`, configurations for touch are commented out, and also `SPI_FREQUENCY` is set to 25MHz as below:

```c++
#define SPI_FREQUENCY       25000000  // 80 MHz
#define SPI_READ_FREQUENCY  25000000  // tested 25000000, 20000000, 5000000
#define SPI_TOUCH_FREQUENCY   250000 // 250 KHz
```

And I should change my conclusion:

- `lcd.init()` prevents SPI for SD.
- `SPI.begin()` for SD and `SD.begin()` prevents LCD working.

----------

There's one thing I noticed. I will report in the new thread.
Anyway, thanks for your help!

**Comment by embedded-kiddie** [comment, marked as answer]
Immediately after compiling and uploading, no output is displayed on the serial monitor.

Therefore, usually, I open the serial monitor, turn the USB switch off and then on to check the message on serial monitor.

In this case, the message "Failed to save to SD" was displayed.

However, when I pressed the reset button on XIAO (which is very small and difficult to press), I found that the SD was working properly.

Now, I open an empty sketch, activate the serial monitor in advance, and then compile and upload the target sketch. The bootloader resets XIAO immediately after uploading, so I can always confirm that the SD is working properly.

![ArduinoIDE](https://github.com/user-attachments/assets/65277908-2c3f-4919-83b8-ca29d4fbbc54)

Therefore, I found that pressing the reset button instead of manipulating the USB switch ensured that the SD always worked properly.

Here is a photo of my testbed. The battery switch acts just like the USB switch whenever I'm not powered by USB. However, this is not an issue as nothing is saved to the SD card immediately after turning on the battery switch.

![MySystem](https://github.com/user-attachments/assets/29d4b5f6-4e5d-4457-affe-fba0982ac6ea)

-----

Based on the above facts and [my test results](#discussion-7171960) where **a white square PNG image (128x128) was saved**,

> And when I tried inserting SPI.begin(...) just before do {...} while (...); as shown below, **a white square PNG image was saved** (I think this means the save failed), but also the message `PNG save success.` was not displayed **ON LCD**.

I can conclude the following:

- Accessing the SD card (`SPI.begin()` for SD or `SD.begin()`) has a negative effect on the LGFX.
- Therefore, `saveToSD()` or `lcd.createPng()` will not work properly and a white square image will be saved.
- Also, displaying on the LCD during `loop()` will not work.
- This issue occurs on ESP32S3 (or at least XIAO ESP32S3) with ST7789.

So I would like to raise this discussion as an issue. Is that okay?
Thank you for your cooperation!

**Reply by embedded-kiddie** [reply]
I also tested with ILI9341 type LCD. And the result is the same as ST7789 🥺
So I should change my conclusion like this:

- This issue occurs on ESP32S3 (or at least XIAO ESP32S3) with ST7789 and ILI9341

By the way, before the test with ILI9341, I executed the [PDQ graphicstest benchmark](https://github.com/embedded-kiddie/Arduino-XIAO-ESP32/tree/main/GFXbenchmark "Arduino-XIAO-ESP32/GFXbenchmark at main · embedded-kiddie/Arduino-XIAO-ESP32") that shows LovyanGFX is still the best!

![PXL_20240913_042003997](https://github.com/user-attachments/assets/0e5d8fbd-eec1-4dc8-a86d-50fb94ea688b)

**Comment by embedded-kiddie** [comment]
This issue was solved at #617 .
